src/test/java/org/jenkinsci/plugins/github_branch_source/EventsTest.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.commons.io.IOUtils;
+import org.jenkinsci.plugins.github.extension.GHSubscriberEvent;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class EventsTest {
+    /**
+     * All tests in this class only use Jenkins for the extensions
+     */
+    @ClassRule
+    public static JenkinsRule r = new JenkinsRule();
+
+    private static SCMEvent.Type firedEventType;
+    private static GHSubscriberEvent ghEvent;
+
+    @Test
+    public void pushEvents() throws Exception {
+        PushGHEventSubscriber subscriber = new PushGHEventSubscriber();
+
+        firedEventType = SCMEvent.Type.CREATED;
+        ghEvent = callOnEvent(subscriber, "EventsTest/pushEventCreated.json");
+        waitAndAssertReceived(true);
+
+        firedEventType = SCMEvent.Type.REMOVED;
+        ghEvent = callOnEvent(subscriber, "EventsTest/pushEventRemoved.json");
+        waitAndAssertReceived(true);
+
+        firedEventType = SCMEvent.Type.UPDATED;
+        ghEvent = callOnEvent(subscriber, "EventsTest/pushEventUpdated.json");
+        waitAndAssertReceived(true);
+    }
+
+    @Test
+    public void pullRequestEvents() throws Exception {
+        PullRequestGHEventSubscriber subscriber = new PullRequestGHEventSubscriber();
+
+        firedEventType = SCMEvent.Type.CREATED;
+        ghEvent = callOnEvent(subscriber, "EventsTest/pullRequestEventCreated.json");
+        waitAndAssertReceived(true);
+
+        firedEventType = SCMEvent.Type.REMOVED;
+        ghEvent = callOnEvent(subscriber, "EventsTest/pullRequestEventRemoved.json");
+        waitAndAssertReceived(true);
+
+        firedEventType = SCMEvent.Type.UPDATED;
+        ghEvent = callOnEvent(subscriber, "EventsTest/pullRequestEventUpdated.json");
+        waitAndAssertReceived(true);
+
+        ghEvent = callOnEvent(subscriber, "EventsTest/pullRequestEventUpdatedSync.json");
+        waitAndAssertReceived(true);
+    }
+
+    @Test
+    public void repositoryEvents() throws Exception {
+        GitHubRepositoryEventSubscriber subscriber = new GitHubRepositoryEventSubscriber();
+
+        firedEventType = SCMEvent.Type.CREATED;
+        ghEvent = callOnEvent(subscriber, "EventsTest/repositoryEventCreated.json");
+        waitAndAssertReceived(true);
+
+        ghEvent = callOnEvent(subscriber, "EventsTest/repositoryEventNotFiredNotFork.json");
+        waitAndAssertReceived(false);
+
+        ghEvent = callOnEvent(subscriber, "EventsTest/repositoryEventNotFiredWrongAction.json");
+        waitAndAssertReceived(false);
+    }
+
+    private GHSubscriberEvent callOnEvent(PushGHEventSubscriber subscriber, String eventPayloadFile) throws IOException {
+        GHSubscriberEvent event = createMockEvent(eventPayloadFile);
```

Comment:
Better to either use `SCMEvent.fireNow(createMockEvent)` given that you have registered as a test extension or submit the payload to the webhook endpoint directly, e.g. `ExtensionList.lookup(RootAction.class).get(com.cloudbees.jenkins.GitHubWebHook.class).doIndex(GHEvent, payload)` otherwise you are not really testing the functionality